src/run.ts

```diff
@@ -0,0 +1,315 @@
+
+    return selectedScenarioNames;
+  }
+}
+
+function updateContext(context: Context, partialContext: PartialContext) {
+  const newContext = {
+    ...context,
+    ...(typeof partialContext === 'function'
+      ? partialContext(context)
+      : partialContext),
+  };
+
+  return newContext;
+}
+
+function mergeMocks(scenarioMap: ({ mocks: Mock[] } | Mock[])[]) {
+  return scenarioMap.reduce<Mock[]>(
+    (result, scenarioMock) =>
+      result.concat(
+        Array.isArray(scenarioMock) ? scenarioMock : scenarioMock.mocks,
+      ),
+    [],
+  );
+}
+
+function getMocksFromScenarios(scenarios: Scenario[]) {
+  const mocks = mergeMocks(scenarios);
+  const httpMocks = getHttpMocks(mocks);
+  const graphQlMocks = getGraphQlMocks(mocks);
+
+  return { httpMocks, graphQlMocks };
+}
+
+function getScenarios({
+  defaultScenario,
+  scenarioMap,
+  scenarioNames,
+}: {
+  defaultScenario: DefaultScenario;
+  scenarioMap: ScenarioMap;
+  scenarioNames: string[];
+}): Scenario[] {
+  return [defaultScenario].concat(
+    scenarioNames.map(scenario => scenarioMap[scenario]),
+  );
+}
+
+function getContextFromScenarios(scenarios: Scenario[]) {
+  let context: Context = {};
+  scenarios.forEach(mock => {
+    if (!Array.isArray(mock) && mock.context) {
+      context = { ...context, ...mock.context };
+    }
+  });
+
+  return context;
+}
+
+function createRequestHandler({
+  getScenarioNames,
+  defaultScenario,
+  scenarioMap,
+  getContext,
+  setContext,
+}: {
+  getScenarioNames: (req: Request) => string[];
+  defaultScenario: DefaultScenario;
+  scenarioMap: ScenarioMap;
+  getContext: (req: Request, selectedScenarios: Scenario[]) => Context;
+  setContext: (res: Response, context: Context) => void;
+}) {
+  return (req: Request, res: Response, next: NextFunction) => {
+    const scenarioNames = getScenarioNames(req);
+    const selectedScenarios = getScenarios({
+      defaultScenario,
+      scenarioMap,
+      scenarioNames,
+    });
+
+    const { httpMocks, graphQlMocks } = getMocksFromScenarios(
+      selectedScenarios,
+    );
+    let context: Context = getContext(req, selectedScenarios);
+
+    const graphQlMock = getGraphQlMock(req, graphQlMocks);
+
+    if (graphQlMock) {
+      const requestHandler = createGraphQlRequestHandler({
+        graphQlMock,
+        updateContext: localUpdateContext,
+        getContext: localGetContext,
+      });
+
+      requestHandler(req, res, next);
+
+      return;
+    }
+
+    const { httpMock, params } = getHttpMockAndParams(req, httpMocks);
+    if (httpMock) {
+      const requestHandler = createHttpRequestHandler({
+        httpMock,
+        params,
+        getContext: localGetContext,
+        updateContext: localUpdateContext,
+      });
+
+      requestHandler(req, res);
+
+      return;
+    }
+
+    // Nothing matched - default 404 from express
+    next();
+
+    function localUpdateContext(partialContext: PartialContext) {
+      context = updateContext(context, partialContext);
```

Comment:
Don't think this is necessary any more, because `setContext` will set it in the server or cookie. And `getContext` called above will get the value from wherever it has been set elsewhere.

Reply:
Added comment to clarify why this is needed at the moment.